server/blob/blob-aes/src/main/java/org/apache/james/blob/aes/AESBlobStoreDAO.java

```diff
@@ -0,0 +1,155 @@
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.blob.aes;
+
+import java.io.ByteArrayInputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.security.GeneralSecurityException;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+
+import javax.crypto.SecretKey;
+import javax.crypto.SecretKeyFactory;
+import javax.crypto.spec.PBEKeySpec;
+
+import org.apache.commons.io.IOUtils;
+import org.apache.james.blob.api.BlobId;
+import org.apache.james.blob.api.BlobStoreDAO;
+import org.apache.james.blob.api.BucketName;
+import org.apache.james.blob.api.ObjectNotFoundException;
+import org.apache.james.blob.api.ObjectStoreIOException;
+import org.reactivestreams.Publisher;
+
+import com.github.fge.lambdas.Throwing;
+import com.google.common.base.Preconditions;
+import com.google.common.io.ByteSource;
+import com.google.crypto.tink.Aead;
+import com.google.crypto.tink.aead.AeadConfig;
+import com.google.crypto.tink.subtle.AesGcmJce;
+
+import reactor.core.publisher.Mono;
+import reactor.core.scheduler.Schedulers;
+
+public class AESBlobStoreDAO implements BlobStoreDAO {
+    private static final byte[] EMPTY_ASSOCIATED_DATA = new byte[0];
+    private static final int PBKDF2_ITERATIONS = 65536;
+    private static final int KEY_SIZE = 256;
+    private static final String SECRET_KEY_FACTORY_ALGORITHM = "PBKDF2WithHmacSHA256";
+
+    private final BlobStoreDAO underlying;
+    private final Aead aead;
+
+    public AESBlobStoreDAO(BlobStoreDAO underlying, CryptoConfig cryptoConfig) {
+        this.underlying = underlying;
+
+        try {
+            AeadConfig.register();
+
+            SecretKey secretKey = deriveKey(cryptoConfig);
+            aead = new AesGcmJce(secretKey.getEncoded());
+        } catch (GeneralSecurityException e) {
+            throw new RuntimeException("Error while starting AESPayloadCodec", e);
+        }
+    }
+
+    private static SecretKey deriveKey(CryptoConfig cryptoConfig) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        byte[] saltBytes = cryptoConfig.salt();
+        SecretKeyFactory skf = SecretKeyFactory.getInstance(SECRET_KEY_FACTORY_ALGORITHM);
+        PBEKeySpec spec = new PBEKeySpec(cryptoConfig.password(), saltBytes, PBKDF2_ITERATIONS, KEY_SIZE);
+        return skf.generateSecret(spec);
+    }
+
+    public byte[] encrypt(byte[] input) {
+        try {
+            return aead.encrypt(input, EMPTY_ASSOCIATED_DATA);
+        } catch (GeneralSecurityException e) {
+            throw new RuntimeException("Unable to build payload for object storage, failed to encrypt", e);
+        }
+    }
+
+    public byte[] decrypt(byte[] ciphertext) throws IOException {
```

Comment:
This PR sure brings back memories :D 

Sadly this implementation still loads the whole blob data in memory ... The corresponding stream based API had issues because of the way Tink is implemented and the bug is still open to this day https://github.com/google/tink/issues/128

Reply:
> Sadly this implementation still loads the whole blob data in memory ... 

I know.

I considered using a reactor pool to manage a collection of `Cypher`s and use `CypherInputStream` in our concurrent environment.

But it is a bit more than "restore what existed before" ie at the point of 3.5.0.

If you have a SHA-1 from which I can copy from, I'd be happy to work with the Piped input stream in the waiting for a better solution...

Reply:
hmm I don have a sha1 but I have this : 
https://github.com/google/tink/issues/128#issuecomment-420893021 and alternatively https://github.com/google/tink/issues/128#issuecomment-433933047 but I'm not convinced it is a better solution as the implementations in these examples starts an unbounded number of threads ... :(

Reply:
Re-found it, last commit contributes an acceptable version.

https://github.com/apache/james-project/pull/342/commits/d7e7ad9e95be36df7b595d417f9dca2c26da5192

My proposal just ends up writing data potentially to disk... Acceptable as we prefer having large data on disk (not to trash the memory) - and encrypted data size is unknown unless we have it in some kind of byte source - as Object Storage would require the size...

Reply:
@jeantil thoughts on this final patch?

Reply:
I like how it aligns with the way incoming emails are processed. I left a couple comments regarding the asymmetry of read and write operations ( not sure if it was deliberate or not).